File: nodejs/tests/signer.test.js
// 单测：对 ../../test-vectors.json 的每个向量复现 base 与 sign。
import { test } from 'node:test';
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, resolve } from 'node:path';

import { buildSignBase, sign, verifyCallback } from '../src/signer.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const vectorsPath = resolve(__dirname, '../../test-vectors.json');
const vectors = JSON.parse(readFileSync(vectorsPath, 'utf8')).vectors;

test('test-vectors.json 已加载且非空', () => {
  assert.ok(Array.isArray(vectors));
  assert.ok(vectors.length >= 10, `expected >=10 vectors, got ${vectors.length}`);
});

for (const v of vectors) {
  test(`vector: ${v.name} — buildSignBase 与服务端逐字节一致`, () => {
    const base = buildSignBase(v.payload, v.secret);
    assert.equal(base, v.base, `base mismatch for ${v.name}`);
  });

  test(`vector: ${v.name} — sign 与标准答案一致`, () => {
    const s = sign(v.payload, v.secret);
    assert.equal(s, v.sign, `sign mismatch for ${v.name}`);
  });
}

// 回调验签：正例（用算出的 sign 装回 payload）+ 篡改一字节反例。
test('verifyCallback 正例：合法签名通过', () => {
  const secret = 'sk_test_0123456789abcdef0123456789abcdef';
  const payload = {
    merchant_no: 'M00000001',
    order_no: 'P202501010001',
    out_order_no: '202501010001',
    amount: 10000,
    currency: 'PHP',
    status: 'success',
    channel_order_no: null, // null 字段不参与签名
    paid_at: '2025-01-05T12:00:00Z',
  };
  payload.sign = sign(payload, secret);
  assert.equal(verifyCallback(payload, secret), true);
});

test('verifyCallback 反例：篡改业务字段一字节后失败', () => {
  const secret = 'sk_test_0123456789abcdef0123456789abcdef';
  const payload = {
    merchant_no: 'M00000001',
    order_no: 'P202501010001',
    amount: 10000,
    currency: 'PHP',
    status: 'success',
  };
  payload.sign = sign(payload, secret);
  // 篡改金额（一处变化）
  const tampered = { ...payload, amount: 10001 };
  assert.equal(verifyCallback(tampered, secret), false);
});

test('verifyCallback 反例：篡改 sign 自身一字节后失败', () => {
  const secret = 'sk_test_0123456789abcdef0123456789abcdef';
  const payload = { merchant_no: 'M00000001', amount: 10000, status: 'success' };
  payload.sign = sign(payload, secret);
  const flipped = payload.sign[0] === '0' ? '1' : '0';
  const tampered = { ...payload, sign: flipped + payload.sign.slice(1) };
  assert.equal(verifyCallback(tampered, secret), false);
});

test('verifyCallback 反例：用错密钥（代付密钥验代收回调）失败', () => {
  const payPayload = { merchant_no: 'M00000001', amount: 10000, status: 'success' };
  payPayload.sign = sign(payPayload, 'secret_pay');
  assert.equal(verifyCallback(payPayload, 'secret_payout'), false);
});

test('verifyCallback 反例：缺 sign 字段直接判否', () => {
  assert.equal(verifyCallback({ amount: 1 }, 'k'), false);
  assert.equal(verifyCallback(null, 'k'), false);
});

File: go/signer_test.go
package openapi

import (
	"bytes"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

// vectorFile 是 test-vectors.json 的结构（只取需要的字段）。
type vectorFile struct {
	Vectors []vector `json:"vectors"`
}

type vector struct {
	Name    string         `json:"name"`
	Desc    string         `json:"desc"`
	Secret  string         `json:"secret"`
	Payload map[string]any `json:"payload"`
	Base    string         `json:"base"`
	Sign    string         `json:"sign"`
}

// loadVectors 读取仓库根的 test-vectors.json（go 子目录的上一级），
// 用 Decoder+UseNumber 解析以保留整数文本形态，避免 float64 把大整数写成 1e+12。
func loadVectors(t *testing.T) []vector {
	t.Helper()
	path := filepath.Join("..", "test-vectors.json")
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("读取向量文件失败 %s: %v", path, err)
	}
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.UseNumber()
	var vf vectorFile
	if err := dec.Decode(&vf); err != nil {
		t.Fatalf("解析向量文件失败: %v", err)
	}
	if len(vf.Vectors) == 0 {
		t.Fatalf("向量文件为空")
	}
	return vf.Vectors
}

// TestVectorsBaseAndSign 对每个向量断言 BuildSignBase==base 且 Sign==sign。
func TestVectorsBaseAndSign(t *testing.T) {
	for _, v := range loadVectors(t) {
		v := v
		t.Run(v.Name, func(t *testing.T) {
			gotBase := BuildSignBase(v.Payload, v.Secret)
			if gotBase != v.Base {
				t.Errorf("base 不匹配\n期望: %s\n实际: %s", v.Base, gotBase)
			}
			gotSign := Sign(v.Payload, v.Secret)
			if gotSign != v.Sign {
				t.Errorf("sign 不匹配\n期望: %s\n实际: %s", v.Sign, gotSign)
			}
		})
	}
}

// TestVerifyCallbackPositive 用向量构造一个带正确 sign 的回调，验签应通过。
func TestVerifyCallbackPositive(t *testing.T) {
	secret := "sk_test_0123456789abcdef0123456789abcdef"
	payload := map[string]any{
		"merchant_no":      "M00000001",
		"order_no":         "P202501010001",
		"out_order_no":     "202501010001",
		"amount":           json.Number("10000"),
		"actual_amount":    json.Number("10000"),
		"currency":         "PHP",
		"status":           "success",
		"channel_order_no": nil, // null 字段不参与签名。
		"paid_at":          "2025-01-01T08:00:00+08:00",
	}
	payload["sign"] = Sign(payload, secret)

	if !VerifyCallback(payload, secret) {
		t.Fatalf("正确签名的回调验签应通过")
	}
}

// TestVerifyCallbackTamperedByte 篡改一个字节后验签应失败（反例）。
func TestVerifyCallbackTamperedByte(t *testing.T) {
	secret := "sk_test_0123456789abcdef0123456789abcdef"
	payload := map[string]any{
		"merchant_no":   "M00000001",
		"payout_no":     "W202501010001",
		"out_payout_no": "WD202501010001",
		"amount":        json.Number("100000"),
		"currency":      "PHP",
		"status":        "success",
		"finished_at":   "2025-01-01T09:00:00+08:00",
	}
	good := Sign(payload, secret)

	// 篡改业务字段一个字节：amount 末位 0→1。
	tampered := cloneParams(payload)
	tampered["amount"] = json.Number("100001")
	tampered["sign"] = good // 沿用旧签名。
	if VerifyCallback(tampered, secret) {
		t.Fatalf("篡改业务字段后验签不应通过")
	}

	// 篡改 sign 本身一个字节。
	bad := flipLastHexByte(good)
	signTampered := cloneParams(payload)
	signTampered["sign"] = bad
	if VerifyCallback(signTampered, secret) {
		t.Fatalf("篡改 sign 后验签不应通过")
	}

	// 缺失 sign 应失败。
	noSign := cloneParams(payload)
	if VerifyCallback(noSign, secret) {
		t.Fatalf("缺失 sign 验签不应通过")
	}
}

// TestVerifyCallbackWrongSecret 用错误密钥验签应失败（代收/代付密钥用错）。
func TestVerifyCallbackWrongSecret(t *testing.T) {
	payCallback := map[string]any{
		"merchant_no": "M00000001",
		"order_no":    "P1",
		"amount":      json.Number("10000"),
		"status":      "success",
	}
	payCallback["sign"] = Sign(payCallback, "secret_pay")

	if VerifyCallback(payCallback, "secret_payout") {
		t.Fatalf("用错误密钥（payout 验代收回调）不应通过")
	}
	if !VerifyCallback(payCallback, "secret_pay") {
		t.Fatalf("用正确密钥应通过")
	}
}

// flipLastHexByte 翻转 hex 串末位字符，制造一字节差异。
func flipLastHexByte(s string) string {
	if s == "" {
		return s
	}
	b := []byte(s)
	last := b[len(b)-1]
	if last == 'f' {
		b[len(b)-1] = 'e'
	} else {
		b[len(b)-1] = last + 1
	}
	return string(b)
}
